### PR TITLE
Duplicated Queries Removed

### PR DESCRIPTION
**Pull request by devansh-webkul**
**BUG:**
Link 1: https://github.com/bagisto/bagisto/issues/4290
Link 2: https://github.com/bagisto/bagisto/issues/4289
Link 3: https://github.com/bagisto/bagisto/issues/4288
Link 4: https://github.com/bagisto/bagisto/issues/4216
Link 5: https://github.com/bagisto/bagisto/issues/4369

**INFO:**

- 98 percent of duplicated queries removed.

- Static loaded config array added to the main entry point of core config method which helps in reducing the duplicated core config values as for fetching core configs this method will be responsible.

- Added core config exception property which gives the ability to ignore those keys which don't want to load in a static array.

### CODE REVIEW

**Comment by vaishaliwebkul** [comment]
@devansh-webkul 
I got some issue with this PR

1. currency exchange rate is not updating ( eg; USD->EUR)
2. customer group pricing is not working

**Comment by vaishaliwebkul** [comment]
@devansh-webkul 
getting exception while adding compare item as guest
https://prnt.sc/vv1jw6

**Comment by vaishaliwebkul** [comment]
fix for the image search option as well.

**Comment by devansh-webkul** [comment]
@vaishaliwebkul,

I confirmed that this bug exists in master. As I haven't done changes in this PR as well.

**Review by vaishaliwebkul — APPROVED** [review]

**Comment by devansh-webkul** [comment]
I found some more duplicated queries in layered navigation.

**Comment by devansh-webkul** [comment]
Marking this as a draft, as need to do some more optimization as well.

File: packages/Webkul/Product/src/Models/Product.php [hunk 1]
```diff
@@ -225,6 +226,14 @@ public function customer_group_prices()
         return $this->hasMany(ProductCustomerGroupPriceProxy::modelClass());
     }
 
+    /**
+     * Get the catalog rule prices.
+     */
+    public function catalog_rule_prices()
```
Comment (jitendra-webkul) [inline packages/Webkul/Product/src/Models/Product.php:232]: Remove this two-way relationship from here as we don't want the Product package to dependent on CatalogRule package

File: packages/Webkul/Core/src/Core.php [hunk 2]
```diff
@@ -276,7 +288,106 @@ public function getAllCustomerGroups()
     }
 
     /**
-     * Returns all currencies
+     * Returns guest group.
+     *
+     * @return object
+     */
+    public function getCustomerGuestGroup()
```
Comment (jitendra-webkul) [inline packages/Webkul/Core/src/Core.php:295]: Move this code to their respective package

File: packages/Webkul/Core/src/Core.php [hunk 3]
```diff
@@ -276,7 +288,106 @@ public function getAllCustomerGroups()
...
+
+        return $customerGuestGroup = $this->customerGroupRepository->findOneByField('code', 'guest');
+    }
+
+    /**
+     * Check if product customer group prices already loaded then load from it.
+     *
+     * @return object
+     */
+    public function checkInLoadedCustomerGroupPrice($product, $customerGroupId)
```
Comment (jitendra-webkul) [inline packages/Webkul/Core/src/Core.php:311]: Move this code to their respective package

File: packages/Webkul/Core/src/Core.php [hunk 4]
```diff
@@ -276,7 +288,106 @@ public function getAllCustomerGroups()
...
+            return $customerGroupPrice->customer_group_id == $customerGroupId || is_null($customerGroupPrice->customer_group_id);
+        });
+    }
+
+    /**
+     * Check if catalog rule prices already loaded. If already loaded then load from it.
+     *
+     * @return object
+     */
+    public function checkInLoadedCatalogRulePrice($product, $customerGroupId)
```
Comment (jitendra-webkul) [inline packages/Webkul/Core/src/Core.php:329]: Move this code to their respective package

File: packages/Webkul/Core/src/Core.php [hunk 5]
```diff
@@ -276,7 +288,106 @@ public function getAllCustomerGroups()
...
+                && $catalogRulePrice->rule_date == Carbon::now()->format('Y-m-d');
+        })->first();
+    }
+
+    /**
+     * Check if product inventories are already loaded. If already loaded then load from it.
+     *
+     * @return object
+     */
+    public function checkInLoadedProductInventories($product)
```
Comment (jitendra-webkul) [inline packages/Webkul/Core/src/Core.php:349]: Move this code to their respective package

File: packages/Webkul/Core/src/Core.php [hunk 6]
```diff
@@ -276,7 +288,106 @@ public function getAllCustomerGroups()
...
+            return $productInventories[$product->id];
+        }
+
+        return $productInventories[$product->id] = $product->inventories;
+    }
+
+    /**
+     * @return object
+     */
+    public function checkInLoadedSaleableChecks($product, $callback)
```
Comment (jitendra-webkul) [inline packages/Webkul/Core/src/Core.php:363]: Move this code to their respective package

File: packages/Webkul/Core/src/Core.php [hunk 7]
```diff
@@ -276,7 +288,106 @@ public function getAllCustomerGroups()
...
+            return $loadedSaleableChecks[$product->id];
+        }
+
+        return $loadedSaleableChecks[$product->id] = $callback($product);
+    }
+
+    /**
+     * @return object
+     */
+    public function checkInLoadedFamilyAttributes($product)
```
Comment (jitendra-webkul) [inline packages/Webkul/Core/src/Core.php:377]: Move this code to their respective package

File: packages/Webkul/Core/src/Core.php [hunk 8]
```diff
@@ -1078,7 +1154,27 @@ public function getSingletonInstance($className)
     }
 
     /**
-     * Returns a string as selector part for identifying elements in views
+     * Returns exchange rates.
+     *
+     * @return collection
+     */
+    public function getChannelInventorySourceIds()
```
Comment (jitendra-webkul) [inline packages/Webkul/Core/src/Core.php:1161]: Move this code to their respective package

File: packages/Webkul/Core/src/Models/Channel.php [hunk 9]
```diff
@@ -68,6 +69,14 @@ public function inventory_sources()
         return $this->belongsToMany(InventorySourceProxy::modelClass(), 'channel_inventory_sources');
     }
 
+    /**
+     * Get the product ordered inventory associated with channel.
+     */
+    public function product_ordered_inventories()
```
Comment (jitendra-webkul) [inline packages/Webkul/Core/src/Models/Channel.php:75]: Remove this two-way relationship from here as we don't want the Core package to dependent on other package

File: packages/Webkul/Product/src/Models/Product.php [hunk 10]
```diff
@@ -381,15 +388,15 @@ public function getCustomAttributeValue($attribute)
 
         if ($attribute->value_per_channel) {
             if ($attribute->value_per_locale) {
-                $attributeValue = $this->attribute_values()->where('channel', $channel)->where('locale', $locale)->where('attribute_id', $attribute->id)->first();
+                $attributeValue = $this->attribute_values->where('channel', $channel)->where('locale', $locale)->where('attribute_id', $attribute->id)->first();
```
Comment (jitendra-webkul) [inline packages/Webkul/Product/src/Models/Product.php:391]: Don't filter on Collection object instead filter on the query builder

File: packages/Webkul/Product/src/Models/Product.php [hunk 11]
```diff
@@ -381,15 +388,15 @@ public function getCustomAttributeValue($attribute)
...
             } else {
-                $attributeValue = $this->attribute_values()->where('channel', $channel)->where('attribute_id', $attribute->id)->first();
+                $attributeValue = $this->attribute_values->where('channel', $channel)->where('attribute_id', $attribute->id)->first();
```
Comment (jitendra-webkul) [inline packages/Webkul/Product/src/Models/Product.php:393]: Don't filter on Collection object instead filter on the query builder

File: packages/Webkul/Product/src/Models/Product.php [hunk 12]
```diff
@@ -381,15 +388,15 @@ public function getCustomAttributeValue($attribute)
...
             }
         } else {
             if ($attribute->value_per_locale) {
-                $attributeValue = $this->attribute_values()->where('locale', $locale)->where('attribute_id', $attribute->id)->first();
+                $attributeValue = $this->attribute_values->where('locale', $locale)->where('attribute_id', $attribute->id)->first();
```
Comment (jitendra-webkul) [inline packages/Webkul/Product/src/Models/Product.php:397]: Don't filter on Collection object instead filter on the query builder

File: packages/Webkul/Product/src/Models/Product.php [hunk 13]
```diff
@@ -381,15 +388,15 @@ public function getCustomAttributeValue($attribute)
...
             } else {
-                $attributeValue = $this->attribute_values()->where('attribute_id', $attribute->id)->first();
+                $attributeValue = $this->attribute_values->where('attribute_id', $attribute->id)->first();
```
Comment (jitendra-webkul) [inline packages/Webkul/Product/src/Models/Product.php:399]: Don't filter on Collection object instead filter on the query builder

**Review by finalbytes — APPROVED** [review]

**Comment by devansh-webkul** [comment]
I am marking this PR as a draft till merging of this PR,
https://github.com/bagisto/bagisto/pull/4550

**Comment by ghermans** [comment]
@devansh-webkul the mentioned PR has been merged.

**Comment by devansh-webkul** [comment]
@ghermans,

Just minor changes also needed, to give Facade support.

**Comment by ghermans** [comment]
@devansh-webkul can you please resolve the conflict of this PR?

**Review by rahulshukla-webkul — APPROVED** [review]

**Pull request merged**